_posts/2020-05-26-introduction-to-spring-webflux.md

```diff
@@ -0,0 +1,170 @@
+
+There was indeed a learning curve with WebFlux and its functional programming style. It took me couple of weeks to get comfortable with it and start seeing the benefits. For example, when porting functionality from the old codebase to our new application, it was quite easy to see problematic patterns or clear architectural violations in the old codebase, as trying to implement similar functionality with WebFlux was surprisingly hard. It was clear quite quick that if you need to use old codebase as a reference, you only check what you need to do (e.g. which calls are needed to backend systems) but not how to actually do it.
+
+### Tips for some common pitfalls you might encounter
+
+#### Returning an empty value
+Sometimes you just need to check something in a function, without returning a value. For example, you might need a function to check if some operation can be made. Your function returns a void value in a success case and throws an exception in an error case:
+```java
+public Foo doSomething() {
+
+  // Get state etc.
+  Bar bar = getBar();
+
+  checkOperationXxx(bar);
+
+  return operationXxx(bar); // Returns foo
+}
+
+void checkOperationXxx(Bar bar) {
+  // Check for access rights, etc.
+  if (bar.isOperationXxxAllowed()) {
+    return;
+  }
+  throw new OperationDeniedException();
+}
+```
+You might try to implement this with WebFlux as follows:
+```java
+public Mono<Foo> doSomething() {
+  return getBar() // Returns Mono<Bar>
+    .map(bar -> checkOperationXxx(bar)
+      .onErrorResume(throwable -> { // Handle error })
+      .map(aValue -> operationXxx(bar)));
+}
+
+Mono<Void> checkOperationXxx(Bar bar) {
+  // Check for access rights, etc.
+  if (bar.isOperationXxxAllowed()) {
+    return Mono.empty();
+  }
+  return Mono.error(new OperationDeniedException());
+}
+```
+However, when you return an empty value with `Mono.empty()` the result really is an empty value. You can't execute any operation such as map or flatMap to it. Your `operationXxx()` is never executed and you'll spend hours trying to figure out what's wrong.
+
+Your next attempt might be to write the caller function as:
+```java
+public Mono<Foo> doSomething() {
+  return getBar() // Returns Mono<Bar>
+    .map(bar -> checkOperationXxx(bar)
+      .onErrorResume(throwable -> { // Handle error })
+      .then(operationXxx(bar)));
+}
+```
+This does work in an success case: `checkOperationXxx()` completes with empty value and then you start executing a new Mono in `then()`-function. However, `operationXxx()` is executed also in an error case! This is not what you want.
```

Comment:
'This does work in case of a success: `checkOperationXxx()` completes with an empty value and you start executing a new `Mono` in the `then()` function. However, `operationXxx()` is also executed in case of an error! This is not what you want.'